### ISSUE TITLE
ISA-Tab representation in PSQL takes up lots of space

### ISSUE THREAD

**Issue by flekschas** [parklab/refinery-platform#843]
After trying to import some ISA-Tabs from MetaboLights I noticed that my Postgres DB grew to 20 GB! In comparison, the total size of all MetaboLight ISA-Tabs together is just 38.2 MB.

Below is an overview of my 20 largest tables:

```
                      relation                      | total_size 
----------------------------------------------------+------------
 public.data_set_manager_annotatednode              | 20 GB
 public.data_set_manager_node                       | 156 MB
 public.data_set_manager_attribute                  | 95 MB
 public.data_set_manager_protocolreference          | 86 MB
 public.data_set_manager_node_children              | 42 MB
 public.data_set_manager_node_parents               | 42 MB
 public.data_set_manager_protocolreferenceparameter | 40 MB
 public.djcelery_taskstate                          | 33 MB
 public.file_store_filestoreitem                    | 20 MB
 public.celery_taskmeta                             | 10 MB
 public.data_set_manager_protocol                   | 7040 kB
 public.data_set_manager_attributeorder             | 4536 kB
 public.data_set_manager_nodecollection             | 1712 kB
 public.guardian_userobjectpermission               | 728 kB
 public.data_set_manager_annotatednoderegistry      | 576 kB
 public.data_set_manager_assay                      | 568 kB
 public.core_datasetuserobjectpermission            | 504 kB
 public.core_dataset                                | 392 kB
 public.core_nodeset                                | 368 kB
 public.data_set_manager_contact                    | 344 kB
```

The query to reproduce the table is:

```
SELECT nspname || '.' || relname AS "relation",
    pg_size_pretty(pg_total_relation_size(C.oid)) AS "total_size"
  FROM pg_class C
  LEFT JOIN pg_namespace N ON (N.oid = C.relnamespace)
  WHERE nspname NOT IN ('pg_catalog', 'information_schema')
    AND C.relkind <> 'i'
    AND nspname !~ '^pg_toast'
  ORDER BY pg_total_relation_size(C.oid) DESC
  LIMIT 20;
```

More here: https://wiki.postgresql.org/wiki/Disk_Usage

Clearly something is wrong with `data_set_manager_annotatednode`.

This problem is probably related to #568.

Total number of annotated nodes: 22.331.943
Distinct* number of annotated nodes: 2.025.396

So roughly 90% if all nodes are duplicates.

* Distinct according to all property except the automatically incremented id.

**Comment by flekschas** [comment]
It turns out that that something is wrong with the nodes and attribute collection when adding annotated nodes. The dataset mentioned above has 533 nodes of 2 file types. Node of the first file type (132) are extended with an annotated node for each attribute (529) making in total 529 * 132 = **69.828** annotated nodes. Now for all remaining nodes of the second file type (401) the number of parent attributes suddenly bumps up to **69.828** which would lead to 69.828 * 401 = 28.001.028 additional annotated nodes. I assume that this is the reason why the import takes forever.

Probably the issue is related to fetching the nodes as well as fetching their parents' attributes.

This method retrieves the nodes and sets the nodes parents:
https://github.com/parklab/refinery-platform/blob/develop/refinery/data_set_manager/utils.py#L395

This method retrieves the parents' attributes:
https://github.com/parklab/refinery-platform/blob/develop/refinery/data_set_manager/utils.py#L407

**Comment by ngehlenborg** [comment]
Good observation. It seems that `_get_parent_attributes(nodes, node_id)` is not doing what it should be doing or that there is an issue with the ISA-Tab file (can you attach it here?). 

I assume that the two file types are raw data files and derived data files, is that correct?

**Comment by flekschas** [comment]
I've attached the file here: https://github.com/parklab/refinery-platform/issues/842#issuecomment-170273812

Yes the two file types are raw data files.

**Comment by ngehlenborg** [comment]
It appears that there are some issues with the assay files in the ISA archive referenced in #842. E.g. the `(Labeled) Extract Name` columns contain `N/A` for all rows and in one of the assay files the normalization and data transformation columns contain a single file name for all rows. This will significantly increase the number of attributes for each "downstream" node, although that was most likely not intended. 

As discussed with @flekschas, we will try to remove all duplicated `node/attribute/attribute value` tuples before passing them on, an return a warning (initially into the log file, in the future directly to the user).

**Comment by flekschas** [comment]
I notice something else: when compiling the list of nodes related to a study and assay, nodes without an assay are always included: https://github.com/parklab/refinery-platform/blob/develop/refinery/data_set_manager/utils.py#L167-L171

Is there a reason for this behavior?

**Comment by ngehlenborg** [comment]
Yes, nodes that are not assigned an assay are part of the study and apply to all assays of that study.

**Comment by flekschas** [comment]
I see. I've tried to import only unique attributes but unfortunately that didn't help. The problematic data set above still caused imports of ~70.000 annotated nodes or more (I stopped the import process) to be imported. I'll investigate further.

**Comment by flekschas** [comment]
Here's some stats from the import process

```
data_set_manager.utils:325 update_annotated_nodes() - 0 annotated nodes deleted.
data_set_manager.utils:338 update_annotated_nodes() - Total number of compiled custom nodes: 533
data_set_manager.utils:342 update_annotated_nodes() - Number of compiled custom nodes of node type Free Induction Decay Data File: 132
data_set_manager.utils:387 update_annotated_nodes() - Skipped creating 34584 duplicated annotated nodes
data_set_manager.utils:390 update_annotated_nodes() - 35244 annotated nodes generated in 16.5440590382
data_set_manager.utils:491 _index_annotated_nodes() - 132 nodes for indexing
```

So the total number of _real_ nodes is 132 and each attribute node is exactly associated with 267 attributes making in total 35244 annotated nodes.

I also notice an error popping up during the import but it doesn't seem to cause any direct harm, i.e. the import process doesn't stop:

```
ERROR    data_set_manager.isa_tab_parser:391 _parse_node() - Unexpected element Acquisition Parameter Data File when parsing node in line 3, column 27.
```

**Comment by hackdna** [comment]
<a href="https:&#x2F;&#x2F;trello.com&#x2F;c&#x2F;uPviGMWp&#x2F;199-improve-data-model-of-data-sets"><img src="https:&#x2F;&#x2F;github.trello.services&#x2F;images&#x2F;trello-icon.png" width="12" height="12"> Improve data model of data sets</a>

**Comment by scottx611x** [comment]
Related to: #1550
